### DISCUSSION TITLE
Problem with connection to my IB paper account via IB Gateway and IBAPI native Python

### DISCUSSION THREAD

**Discussion by oskoa** [gnzsnz/ib-gateway-docker#126] [Q&A] [answered]
Hi everyone,

I am trying to run IB Gateway on a cloud server (without GUI) and connect to it via python (ibapi) on the same server that IB Gateway is running. 

I have cloned the latest version of ib-gateway-docker and followed instructions on how to build and run with `docker compose up`. 
based on my observation, it looks like the container does run but faces a problems when I try to connect via python.

Here's my simple python snippet for testing the connection:
```
from ibapi.client import EClient
from ibapi.wrapper import EWrapper  

class IBapi(EWrapper, EClient):
     def __init__(self):
         EClient.__init__(self, self) 

app = IBapi()
app.connect('127.0.0.1', 4002, 10)
app.run()

app.disconnect()
```

Here's the container log when I try to run the python script:

```
ib-gateway-1  | 2024-06-24 15:04:42:963 IBC: Login attempt: 1
ib-gateway-1  | 2024-06-24 15:04:42:968 IBC: trading mode from settings: tradingMode=paper
ib-gateway-1  | 2024-06-24 15:04:43:043 IBC: Click button: Paper Log In
ib-gateway-1  | 2024-06-24 15:04:45:503 IBC: detected frame entitled: IBKR Gateway; event=Lost focus
ib-gateway-1  | 2024-06-24 15:04:45:504 IBC: detected frame entitled: IBKR Gateway; event=Deactivated
ib-gateway-1  | 2024-06-24 15:04:45:514 IBC: detected frame entitled: Loading...; event=Activated
ib-gateway-1  | 2024-06-24 15:04:45:545 IBC: detected frame entitled: Loading...; event=Focused
ib-gateway-1  | 2024-06-24 15:04:45:546 IBC: detected frame entitled: Loading...; event=Opened
ib-gateway-1  | 2024-06-24 15:04:45:547 IBC: Found Gateway main window
ib-gateway-1  | 2024-06-24 15:04:45:548 IBC: Got main window from future
ib-gateway-1  | 2024-06-24 15:04:45:548 IBC: Invoking config dialog menu
ib-gateway-1  | 2024-06-24 15:04:46:099 IBC: detected frame entitled: Loading...; event=Lost focus
ib-gateway-1  | 2024-06-24 15:04:46:101 IBC: detected frame entitled: Loading...; event=Deactivated
ib-gateway-1  | 2024-06-24 15:04:46:103 IBC: detected frame entitled: Authenticating...; event=Activated
ib-gateway-1  | 2024-06-24 15:04:46:109 IBC: detected frame entitled: Connecting to server...; event=Focused
ib-gateway-1  | 2024-06-24 15:04:46:110 IBC: detected frame entitled: Connecting to server...; event=Opened
ib-gateway-1  | 2024-06-24 15:04:47:369 IBC: detected frame entitled: Authenticating...; event=Lost focus
ib-gateway-1  | 2024-06-24 15:04:47:370 IBC: detected frame entitled: Authenticating...; event=Deactivated
ib-gateway-1  | 2024-06-24 15:04:47:372 IBC: detected dialog entitled: Enter security code; event=Activated
ib-gateway-1  | 2024-06-24 15:04:47:386 IBC: detected dialog entitled: Enter security code; event=Focused
ib-gateway-1  | 2024-06-24 15:04:47:387 IBC: detected dialog entitled: Enter security code; event=Opened
ib-gateway-1  | 2024/06/24 15:06:14 socat[146] E connect(5, AF=2 127.0.0.1:4002, 16): Connection refused
ib-gateway-1  | 2024/06/24 15:07:45 socat[147] E connect(5, AF=2 127.0.0.1:4002, 16): Connection refused
ib-gateway-1  | 2024-06-24 15:07:49:509 IBC: detected dialog entitled: Enter security code; event=Closed
ib-gateway-1  | 2024/06/24 15:08:46 socat[153] E connect(5, AF=2 127.0.0.1:4002, 16): Connection refused
```

It looks like 2FA is a problem (Enter security code) . The python code triggers a text message to my phone. Does anyone know how to deal with 2FA? I doubt I have deactivate it since it is required every time I login to my paper account via the GUI. 

I am attaching my .env for reference:

```
TWS_USERID=MyUserID
TWS_PASSWORD=MyPass
# ib-gateway
#TWS_SETTINGS_PATH=/home/ibgateway/Jts
# tws
#TWS_SETTINGS_PATH=/config/tws_settings
TWS_SETTINGS_PATH=
TWS_ACCEPT_INCOMING=
TRADING_MODE=paper
READ_ONLY_API=no
VNC_SERVER_PASSWORD=
TWOFA_TIMEOUT_ACTION=restart
BYPASS_WARNING=
AUTO_RESTART_TIME=11:59 PM
AUTO_LOGOFF_TIME=
TWS_COLD_RESTART=
SAVE_TWS_SETTINGS=
RELOGIN_AFTER_2FA_TIMEOUT=yes
TIME_ZONE=Europe/Zurich
CUSTOM_CONFIG=
SSH_TUNNEL=
SSH_OPTIONS=
SSH_ALIVE_INTERVAL=
SSH_ALIVE_COUNT=
SSH_PASSPHRASE=
SSH_REMOTE_PORT=
SSH_USER_TUNNEL=
SSH_RESTART=
SSH_VNC_PORT=
```
 

I do not modify the yml file at all and use what is there on the repo. 

I appreciate your help with this.

**Comment by gnzsnz** [comment]
I would suggest you to connect through VNC to make sure everything is working fine. ideally you don't need it in "steady mode" but the main reason to have it is to solve any issue.

i don't understand why you have MFA for paper, I don't have it for my paper account, and AFAIK it's only available for live account. if you have MFA set with text messages, then you need to enter the message manually. it's better to use the ibkr app for MFA as you approve the login in the app.

make sure you can see in the log `IBC: Logging has completed` the API won't work until you complete authentication.

**Reply by oskoa** [reply]
`IBC: Logging has completed` never shows up. The tail of the log is the one I copied above. I believe, aside from the 2FA, there may be other issues?

I am using the IB mobile app but still receive text messages somehow. I will look into changing the settings. I was also surprised why I need 2FA for a demo account. 

I will try to update this thread if I find any solution.

**Reply by GonzRon** [reply]
As of February 2025, it is not possible to disable two-factor authentication (2FA) for Interactive Brokers (IB) paper trading accounts. Here's what you need to know:

1. 2FA is mandatory for all IB accounts, including paper trading accounts[1][2].

2. In the past, it was possible to have a second account without 2FA, but this option is no longer available[1].

3. For live trading accounts, some users have reported success in disabling 2FA by going to User settings > Secure login System > Disable two-factor authentication[3]. However, this option is not available for paper trading accounts.

4. If you're experiencing issues with authentication for paper trading, ensure that:
   - You're not logged in to the IB website simultaneously
   - You're attempting to log in during trading hours
   - Your account is upgraded from Lite to Pro (this may take a day to process)[2]

5. It's important to note that 2FA is considered a basic security practice for important accounts, and it's generally recommended to keep it enabled for enhanced account protection[3].

If you're still experiencing issues with your paper trading account authentication, it's best to contact Interactive Brokers customer support directly for assistance, as they may have the most up-to-date information on account security settings and authentication procedures.

Citations:
[1] https://groups.io/g/ibcalpha/topic/2_factor_authentication_and/4345265
[2] https://www.quantconnect.com/forum/discussion/11963/unable-to-authenticate-with-ib-paper-trading/
[3] https://www.reddit.com/r/interactivebrokers/comments/jub3p0/how_to_disable_ib_key_two_factor_authentication/
[4] https://ibkrguides.com/securelogin/sls/faq.htm
[5] https://www.youtube.com/watch?v=1-y8_23URJ4
[6] https://www.interactivebrokers.com/campus/trading-lessons/ib-key-two-factor-authentication-iphone/
[7] https://www.interactivebrokers.com/campus/trading-lessons/ib-key-two-factor-authentication-android/
[8] https://www.interactivebrokers.com/en/general/secure-login.php

**Reply by gnzsnz** [reply]
I'm using my paper account with no M2F. That's not the case for live account, which request M2F. I haven't found in any of the sources provided anything clearly stating that M2F is required for paper

**Reply by billching123** [reply]
I don't have 2f auth for paper trading.  I am able to login fine with the gui, but with the docker latest and stable from gnzsnz/ib-gateway, it never logs in.  My credentials are in the IBcontroller.ini file.

ib-gateway_1    |     FIXLoginId=***
ib-gateway_1    |     IbLoginId=***
ib-gateway_1    |     LoginDialogDisplayTimeout=60
ib-gateway_1    |     ReadOnlyLogin=no
ib-gateway_1    |     ReloginAfterSecondFactorAuthenticationTimeout=
ib-gateway_1    | 2025-06-11 17:31:19:575 IBC: using default login manager: will get username and password from settings; FIX username and password (if required) from settings
ib-gateway_1    | 2025-06-11 17:31:19:619 IBC: Starting session: will exit if login dialog is not displayed within 60 seconds
ib-gateway_1    | 2025-06-11 17:31:21:815 IBC: Login dialog WINDOW_OPENED: LoginState is LOGGED_OUT

Any help would be appreciated.

**Reply by gnzsnz** [reply]
@billching123 there is no `IBcontroller.ini` file. Please follow instructions on README, use sample `docker-compose.yml` and sample `.env` (on the README) that should work.

**Reply by Lcstyle** [reply]
as a follow up - IB no longer requires 2FA for paper accounts at all, as per their support, and there's no way to enable it or remove from the Paper Portal.  i think there is a way to enable IBKey or the Challenge Response via their Mobile app, though I haven't tried it.

**Comment by dreams-and-thoughts** [comment]
I'm not sure how related this is, but I'm going to put a comment anyway because this discussion came up when I was searching for the following error (Also in the OP logs):

`socat[2381] E connect(5, AF=2 127.0.0.1:7497, 16): Connection refused`

Assuming TWS is logging in correctly, and otherwise working fine, this error for me was related to the setting in TWS  menu File > Global Configuration... > API Settings > Enable ActiveX and Socket Clients

**Reply by boheyse** [reply]
Is there a way we can set this using this docker image?

**Comment by billching123** [comment, marked as answer]
Yes, thank you.  I got it working now.
